lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Layer.hs

```diff
@@ -0,0 +1,260 @@
+    , bfGetLatestEpochProtocolParams = nt bfGetLatestEpochProtocolParams
+    , bfGetPoolHistory = ((nt .) .) . bfGetPoolHistory
+    , bfGetLedgerGenesis = nt bfGetLedgerGenesis
+    , bfGetAccount = nt . bfGetAccount
+    , bfGetBlockSlot = nt . bfGetBlockSlot
+    , bfGetBlockAtHeight = nt . bfGetBlockAtHeight
+    , bfGetBlockAfterHash = nt . bfGetBlockAfterHash
+    , bfGetBlocksAfterHash = nt . bfGetBlocksAfterHash
+    , bfGetBlockTxs = nt . bfGetBlockTxs
+    , bfGetTx = nt . bfGetTx
+    , bfGetTxStakes = nt . bfGetTxStakes
+    , bfGetTxUtxos = nt . bfGetTxUtxos
+    , bfGetTxWithdrawals = nt . bfGetTxWithdrawals
+    , bfGetTxDelegations = nt . bfGetTxDelegations
+    , bfGetTxMetadataJSON = nt . bfGetTxMetadataJSON
+    , bfGetAddressTransactions = ((nt .) .) . bfGetAddressTransactions
+    , bfGetAccountRegistrations = nt . bfGetAccountRegistrations
+    , bfGetAccountDelegations = nt . bfGetAccountDelegations
+    , bfGetAccountWithdrawals = nt . bfGetAccountWithdrawals
+    , bfGetNetworkInfo = nt bfGetNetworkInfo
+    , bfListPools = nt bfListPools
+    }
+
+rateLimitedBlockfrostLayer :: Tracer IO Log -> Project -> IO (BlockfrostLayer IO)
+rateLimitedBlockfrostLayer tr project = do
+    bfConfig <- BFM.newClientConfig project
+    r <- newIORef . (,rps) =<< getPOSIXTime
+    pure $ hoistBlockfrostLayer blockfrostLayer \bfm -> do
+        (prev, n) <- readIORef r
+        time <- getPOSIXTime
+        let remaining = if floorSec time /= floorSec prev then rps else n
+        when (remaining < 1) do
+            let pause = toMicros $ nextSecond time - time
+            traceWith tr $ RateLimiting pause
+            threadDelay pause
+        res <- BFM.run bfConfig bfm
+        writeIORef r (time, remaining - 1)
+        pure res
+  where
+    rps :: Int
+    rps = 8
+
+    floorSec :: POSIXTime -> Integer
+    floorSec = floor . nominalDiffTimeToSeconds
+
+    toMicros :: POSIXTime -> Int
+    toMicros = truncate . (* 1_000_000) . nominalDiffTimeToSeconds
+
+    nextSecond :: POSIXTime -> POSIXTime
+    nextSecond =
+        fromIntegral @Integer . truncate . (+ 1) . nominalDiffTimeToSeconds
+
+newtype Log = RateLimiting Int
+
+instance ToText Log where
+    toText = \case
+        RateLimiting micros ->
+            "Pausing for " <> toText (show micros) <> " microseconds..."
+
+instance HasSeverityAnnotation Log where
+    getSeverityAnnotation = \case RateLimiting{} -> Debug
+
+-- | Query all results, until we get less than maximum items per page.
+-- Until https://github.com/blockfrost/blockfrost-haskell/issues/26 is fixed.
+allPages' :: Monad m => (Paged -> m [a]) -> m [a]
+allPages' act = do
+    let fetch page' = do
+            res <- act page'
+            case res of
+                xs | length xs < maxPageSize -> pure xs
+                xs -> do
+                    next <- fetch (nextPage page')
+                    pure $ xs <> next
+    fetch Paged { countPerPage = maxPageSize, pageNumber = 1 }
```

Comment:
```suggestion
    fetch Paged{ countPerPage = maxPageSize, pageNumber = 1 }
```

My beloved whitespace used to distinguish function arguments. 🤓

Reply:
I'am getting used to it gradually :)